frontend/app/(settings)/admin/locations/building-form.tsx

```diff
@@ -24,44 +26,120 @@ interface RoomFormProps {
   });
+
+  export default function BuildingForm({currentBuilding, closeDialog, refreshTable, createMode = false}: RoomFormProps) {
+  const {sid} = useUser();
+  const [locationType, setLocationType] = useState("ossLink");
+  const useOssLink = locationType === "ossLink";
+  
+  const generateOSSLink = (latitude: number, longitude: number, zoomLevel: number) => {
+    if (latitude === undefined || longitude === undefined || zoomLevel === undefined) { 
+      return "";
+    } 
+    return `https://www.openstreetmap.org/#map=${zoomLevel}/${latitude}/${longitude}`;
+  };
+  const buildingFormSchema = useOssLink ? ossLinkFormSchema : coordinatesFormSchema;
   const form = useForm<z.infer<typeof buildingFormSchema>>({
     resolver: zodResolver(buildingFormSchema),
-    defaultValues: {
+    defaultValues: getFormDefaults(),
+  });
+   function getFormDefaults() {
+    const baseDefaults = {
       name: createMode ? "" : currentBuilding.name,
-      street: createMode ? "": currentBuilding.street,
+      street: createMode ? "" : currentBuilding.street,
       number: createMode ? "" : currentBuilding.number,
       city: createMode ? "" : currentBuilding.city,
       zip: createMode ? "" : currentBuilding.zip,
-      latitude: createMode ? undefined : currentBuilding.latitude,
-      longitude: createMode ? undefined : currentBuilding.longitude,
-      zoomLevel: createMode ? undefined : currentBuilding.zoomLevel
-    },
+    };
+    if (useOssLink) {
+      const ossLink = (!createMode && currentBuilding.latitude && currentBuilding.longitude)
+        ? generateOSSLink(currentBuilding.latitude, currentBuilding.longitude, currentBuilding.zoomLevel)
+        : "";
+      return {...baseDefaults, ossLink};
+    } else {
+      return {
+        ...baseDefaults,
+        latitude: createMode ? undefined : currentBuilding.latitude,
+        longitude: createMode ? undefined : currentBuilding.longitude,
+        zoomLevel: createMode ? undefined : currentBuilding.zoomLevel,
+      };
+    }
+  }
+  const [hasTriedToSubmit, setHasTriedToSubmit] = useState(false);
 
-  });
+useEffect(() => {
+    form.reset(getFormDefaults());
+    setHasTriedToSubmit(false);
+  }, [locationType, createMode, currentBuilding]);
 
-  const [hasTriedToSubmit, setHasTriedToSubmit] = useState(false);
+  function getCoordinatesFromOssLink(link: string): { latitude?: number; longitude?: number; zoomLevel?: number } {
+    if (!link) return {};
 
-  async function onValidSubmit(buildingData: z.infer<typeof buildingFormSchema>) {
+    const match = link.match(/#map=(\d+\.?\d*)\/(\d+\.?\d*)\/(\d+\.?\d*)/);
+    if (match) {
+      return {
+        zoomLevel: parseFloat(match[1]),
+        latitude: parseFloat(match[2]),
+        longitude: parseFloat(match[3]),
+      };
+    }
+    return {};
+  }
+
+async function onValidSubmit(formData: z.infer<typeof buildingFormSchema>) {
     const client = getClient(String(sid));
+    
+    let buildingData: any = {

```

Comment:
```suggestion
    const buildingData = {
```

This can be a const as we only edit the objects fields. Also don't explicitly `any` data, as this beats the purpose of using typescript